<a href="https://colab.research.google.com/github/sujin-create/finedu/blob/main/%EA%B8%88%EC%9C%B5%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%B6%84%EC%84%9D/%EC%A2%85%EB%AA%A9_%EB%89%B4%EC%8A%A4%EC%97%90_%EB%8C%80%ED%95%9C_Word2Vec.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. 파일 로드

In [1]:
!wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1Xlv8I3Q-1fSPDTB6xkA4jSy9Pwj-AOb3' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1Xlv8I3Q-1fSPDTB6xkA4jSy9Pwj-AOb3" -O daily_naver_news.csv && rm -rf /tmp/cookies.txt

--2022-07-23 06:11:05--  https://docs.google.com/uc?export=download&confirm=t&id=1Xlv8I3Q-1fSPDTB6xkA4jSy9Pwj-AOb3
Resolving docs.google.com (docs.google.com)... 74.125.195.138, 74.125.195.102, 74.125.195.100, ...
Connecting to docs.google.com (docs.google.com)|74.125.195.138|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-00-7k-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/jfq0knac5amfks0jh2bnusuma0jp9iep/1658556600000/17609157229046208934/*/1Xlv8I3Q-1fSPDTB6xkA4jSy9Pwj-AOb3?e=download&uuid=67b26a4f-eba4-4a21-8f62-ad7ef8962ad6 [following]
--2022-07-23 06:11:06--  https://doc-00-7k-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/jfq0knac5amfks0jh2bnusuma0jp9iep/1658556600000/17609157229046208934/*/1Xlv8I3Q-1fSPDTB6xkA4jSy9Pwj-AOb3?e=download&uuid=67b26a4f-eba4-4a21-8f62-ad7ef8962ad6
Resolving doc-00-7k-docs.googleusercontent.com (doc-00-7k-docs.googleusercontent.com)... 172.253.117.132,

In [2]:
import pandas as pd
df = pd.read_csv('/content/daily_naver_news.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [3]:
df

name  symbol             datetime  \
0          넷마블  251270  2021-07-19 08:57:00   
1         삼성전자  005930  2021-04-08 06:00:00   
2         삼성전자  005930  2021-05-12 13:40:00   
3          TYM  002900  2021-08-24 15:51:00   
4       LG유플러스  032640  2020-09-24 09:46:00   
...        ...     ...                  ...   
136959    고려아연   10130  2020-07-29 12:01:00   
136960    고려아연   10130  2020-07-27 17:25:00   
136961  하나금융지주   86790  2021-04-07 11:24:00   
136962  현대글로비스   86280  2021-02-19 17:18:00   
136963   카카오뱅크  323410  2021-08-09 18:09:00   

                                             title  \
0             " 넷마블, 하반기 신작 출시 및 관계사 IPO기대···목표가↑"   
1                        "'10만전자' 가려면…" 삼성 실적 변수는?   
2       "'7만전자'까지 밀렸다" 힘 못쓰는 삼성전자…외국인 2조 매도 폭탄 ...   
3          "'K농기계' 6개월 만에 매출 1조원 돌파"…美 사로잡은 대동·TYM   
4                      "'U+ 알뜰폰' 가입자 작년보다 2배이상 늘어"   
...                                            ...   
136959         金 이어 銀도 천장 뚫을 기세… 2분기 부진에도 고려아연 웃는다   
136960                    金·銀값 강세 지속에…고려아연·영풍 '들썩'   
136961             女 골프 샛별 타바타나킷 깜짝 우승에 하나금융도 ‘잭팟’   
136962         女사외이사 늘리는 현대차그룹···기아·모비스·글로비스 선임 예고   
136963                  年3.6%…시중銀보다 높은 카카오뱅크 마통 이자   

                                                  content  \
0                                                     NaN   
1       \n\t\t\t [머니투데이 심재현 기자] '2분기는 더 좋을 것', '위기에서 빛...   
2       \n[이미지출처=연합뉴스][아시아경제 이선애 기자] 외국인의 매도 폭탄으로 국내 증...   
3                                                     NaN   
4       \nLG유플러스 지원들이 편의점에 진열된 알뜰폰 유심을 살피고 있다. [파이낸셜뉴스...   
...                                                   ...   
136959  \n\t\t\t 비철금속 제련업체 고려아연(010130)이 올해 2분기 실적 부진에...   
136960  \n금선물 9년 만에 최고가산업용 비중 높은 은선물 1주일 새 18% 가파른 상승고...   
136961  \n아시아·신남방 지역 홍보 위해 태국 유망주 물색지난해 후원 이후 성적 저조했지만...   
136962  \n기아, 조화순 연세대 교수 추천할듯현대모비스는 강진아 교수 추천 예고글로비스도 ...   
136963                                                NaN   

                                                     link  
0       /item/news_read.nhn?article_id=0003938322&offi...  
1       /item/news_read.nhn?article_id=0004569676&offi...  
2       /item/news_read.nhn?article_id=0004901497&offi...  
3       /item/news_read.nhn?article_id=0004635496&offi...  
4       /item/news_read.nhn?article_id=0004500298&offi...  
...                                                   ...  
136959  /item/news_read.nhn?article_id=0000562271&offi...  
136960  /item/news_read.nhn?article_id=0004388015&offi...  
136961  /item/news_read.nhn?article_id=0000697296&offi...  
136962  /item/news_read.nhn?article_id=0003873556&offi...  
136963  /item/news_read.nhn?article_id=0004589550&offi...  

[136964 rows x 6 columns]

In [4]:
# 종목의 개수
df['name'].nunique()

2403

In [5]:
df.datetime.max()

'2021-09-15 11:55:00'

In [6]:
df.datetime.min()

'2020-05-04 00:01:00'

2403개의 종목에 대해서 2020년 5월 4일부터 2021년 9월 15일까지의 종목 뉴스를 수집하였습니다.

#2. 형태소 분석기 Mecab 설치

In [7]:
!git clone https://github.com/SOMJANG/Mecab-ko-for-Google-Colab.git
%cd Mecab-ko-for-Google-Colab
!bash install_mecab-ko_on_colab190912.sh

Cloning into 'Mecab-ko-for-Google-Colab'...
remote: Enumerating objects: 115, done.
remote: Counting objects: 100% (24/24), done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 115 (delta 11), reused 10 (delta 3), pack-reused 91
Receiving objects: 100% (115/115), 1.27 MiB | 24.09 MiB/s, done.
Resolving deltas: 100% (50/50), done.
/content/Mecab-ko-for-Google-Colab
Installing konlpy.....
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 19.4 MB 1.2 MB/s 
     |████████████████████████████████| 453 kB 55.9 MB/s 
Done
Installing mecab-0.996-ko-0.9.2.tar.gz.....
from https://bitbucket.org/eunjeon/mecab-ko/downloads/mecab-0.996-ko-0.9.2.tar.gz
--2022-07-23 06:12:18--  https://bitbucket.org/eunjeon/mecab-ko/downloads/mecab-0.996-ko-0.9.2.tar.gz
Resolving bitbucket.org (bitbucket.org)... 104.192.141.1, 2406:da00:ff00::3403:4be7, 2406:da00:ff00::6b17:d1f5, ...
Connecting to bitbucket.org (b

#3. Word2Vec

In [8]:
from konlpy.tag import Mecab

In [9]:
tokenizer = Mecab()

In [10]:
# 정규 표현식을 통해서 한국어 제외 문자들을 제거
df['content'] = df['content'].str.replace("[^ㄱ-ㅎㅏ-ㅣ가-힣 ]","")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  


In [11]:
df

name  symbol             datetime  \
0          넷마블  251270  2021-07-19 08:57:00   
1         삼성전자  005930  2021-04-08 06:00:00   
2         삼성전자  005930  2021-05-12 13:40:00   
3          TYM  002900  2021-08-24 15:51:00   
4       LG유플러스  032640  2020-09-24 09:46:00   
...        ...     ...                  ...   
136959    고려아연   10130  2020-07-29 12:01:00   
136960    고려아연   10130  2020-07-27 17:25:00   
136961  하나금융지주   86790  2021-04-07 11:24:00   
136962  현대글로비스   86280  2021-02-19 17:18:00   
136963   카카오뱅크  323410  2021-08-09 18:09:00   

                                             title  \
0             " 넷마블, 하반기 신작 출시 및 관계사 IPO기대···목표가↑"   
1                        "'10만전자' 가려면…" 삼성 실적 변수는?   
2       "'7만전자'까지 밀렸다" 힘 못쓰는 삼성전자…외국인 2조 매도 폭탄 ...   
3          "'K농기계' 6개월 만에 매출 1조원 돌파"…美 사로잡은 대동·TYM   
4                      "'U+ 알뜰폰' 가입자 작년보다 2배이상 늘어"   
...                                            ...   
136959         金 이어 銀도 천장 뚫을 기세… 2분기 부진에도 고려아연 웃는다   
136960                    金·銀값 강세 지속에…고려아연·영풍 '들썩'   
136961             女 골프 샛별 타바타나킷 깜짝 우승에 하나금융도 ‘잭팟’   
136962         女사외이사 늘리는 현대차그룹···기아·모비스·글로비스 선임 예고   
136963                  年3.6%…시중銀보다 높은 카카오뱅크 마통 이자   

                                                  content  \
0                                                     NaN   
1        머니투데이 심재현 기자 분기는 더 좋을 것 위기에서 빛나는 사업 경쟁력일 삼성전자...   
2       이미지출처연합뉴스아시아경제 이선애 기자 외국인의 매도 폭탄으로 국내 증시가 상승 반...   
3                                                     NaN   
4       유플러스 지원들이 편의점에 진열된 알뜰폰 유심을 살피고 있다 파이낸셜뉴스 유플러스는...   
...                                                   ...   
136959   비철금속 제련업체 고려아연이 올해 분기 실적 부진에도 금은 등 귀금속 가격 상승에...   
136960  금선물 년 만에 최고가산업용 비중 높은 은선물 주일 새  가파른 상승고려아연  매출...   
136961  아시아신남방 지역 홍보 위해 태국 유망주 물색지난해 후원 이후 성적 저조했지만 첫 ...   
136962  기아 조화순 연세대 교수 추천할듯현대모비스는 강진아 교수 추천 예고글로비스도 윤윤진...   
136963                                                NaN   

                                                     link  
0       /item/news_read.nhn?article_id=0003938322&offi...  
1       /item/news_read.nhn?article_id=0004569676&offi...  
2       /item/news_read.nhn?article_id=0004901497&offi...  
3       /item/news_read.nhn?article_id=0004635496&offi...  
4       /item/news_read.nhn?article_id=0004500298&offi...  
...                                                   ...  
136959  /item/news_read.nhn?article_id=0000562271&offi...  
136960  /item/news_read.nhn?article_id=0004388015&offi...  
136961  /item/news_read.nhn?article_id=0000697296&offi...  
136962  /item/news_read.nhn?article_id=0003873556&offi...  
136963  /item/news_read.nhn?article_id=0004589550&offi...  

[136964 rows x 6 columns]

In [12]:
news_data = df.content

In [13]:
news_data

0                                                       NaN
1          머니투데이 심재현 기자 분기는 더 좋을 것 위기에서 빛나는 사업 경쟁력일 삼성전자...
2         이미지출처연합뉴스아시아경제 이선애 기자 외국인의 매도 폭탄으로 국내 증시가 상승 반...
3                                                       NaN
4         유플러스 지원들이 편의점에 진열된 알뜰폰 유심을 살피고 있다 파이낸셜뉴스 유플러스는...
                                ...                        
136959     비철금속 제련업체 고려아연이 올해 분기 실적 부진에도 금은 등 귀금속 가격 상승에...
136960    금선물 년 만에 최고가산업용 비중 높은 은선물 주일 새  가파른 상승고려아연  매출...
136961    아시아신남방 지역 홍보 위해 태국 유망주 물색지난해 후원 이후 성적 저조했지만 첫 ...
136962    기아 조화순 연세대 교수 추천할듯현대모비스는 강진아 교수 추천 예고글로비스도 윤윤진...
136963                                                  NaN
Name: content, Length: 136964, dtype: object

In [14]:
# 결측값 수 카운트
news_data.isnull().sum()

34416

In [15]:
# 결측값 삭제
news_data.dropna(inplace=True)

In [16]:
len(news_data)

102548

In [17]:
# 다소 시간 소요
result = news_data.apply(tokenizer.morphs)

In [ ]:
from gensim.models import Word2Vec
model = Word2Vec(result, size=100, window=5, min_count=5, workers=4, sg=0) #sg -> weight의 방향을 반대로 하는 것

윈도우 사이즈를 크게 주면 -> 자주 같이 등장하는 단어들이 유사해짐.  
윈도우 사이즈를 작게 주면 -> 이 단어를 교체할 수 있는 단어들이 유사해짐. (반의어 포함)

In [ ]:
model.wv.most_similar('삼성전자')

[('하이닉스', 0.8553911447525024),
 ('삼성전기', 0.7357660531997681),
 ('디스플레이', 0.68097984790802),
 ('전자', 0.6278015375137329),
 ('현대모비스', 0.6052756905555725),
 ('도시바', 0.5624388456344604),
 ('삼성', 0.5513879656791687),
 ('에릭슨', 0.5393810868263245),
 ('반도체', 0.5342387557029724),
 ('인텔', 0.5295099020004272)]

In [ ]:
model.wv.most_similar('제약')

[('종근당', 0.5766382217407227),
 ('유한양행', 0.5514268279075623),
 ('서창우', 0.5424292087554932),
 ('약품', 0.5169521570205688),
 ('파마', 0.5036357045173645),
 ('동화약품', 0.48121124505996704),
 ('의약품', 0.4746244549751282),
 ('자안', 0.4718179404735565),
 ('휴', 0.467726469039917),
 ('얀센', 0.46495768427848816)]

In [ ]:
# 개수를 지정 가능
model.wv.most_similar('삼성전자', topn=20)

[('하이닉스', 0.8553911447525024),
 ('삼성전기', 0.7357660531997681),
 ('디스플레이', 0.68097984790802),
 ('전자', 0.6278015375137329),
 ('현대모비스', 0.6052756905555725),
 ('도시바', 0.5624388456344604),
 ('삼성', 0.5513879656791687),
 ('에릭슨', 0.5393810868263245),
 ('반도체', 0.5342387557029724),
 ('인텔', 0.5295099020004272),
 ('파운드리', 0.5292571783065796),
 ('하이텍', 0.512376070022583),
 ('만도', 0.5094554424285889),
 ('갤럭시', 0.5082420706748962),
 ('휴켐스', 0.5023633241653442),
 ('테슬라', 0.4979436993598938),
 ('현대차', 0.4963374733924866),
 ('노키아', 0.48208534717559814),
 ('호텔신라', 0.4760553240776062),
 ('한국전력공사', 0.4755811095237732)]

* postive=['A', 'B'], negative=['C']  

'A'와 'C' 정도의 관계를 가지는
'B'의 파트너는 뭐야?

In [ ]:
model.wv.most_similar(positive=['삼성전자', '제약'], negative=['스마트폰'])

[('종근당', 0.5840221047401428),
 ('에이치엘비', 0.5442877411842346),
 ('녹십자', 0.5297243595123291),
 ('사이언스', 0.5211255550384521),
 ('오리엔트', 0.5150229930877686),
 ('하이닉스', 0.5115798115730286),
 ('유한양행', 0.5058728456497192),
 ('대한전선', 0.502576470375061),
 ('로직스', 0.5000935196876526),
 ('파마', 0.49249228835105896)]

In [ ]:
model.wv.most_similar(positive=['삼성전자', '자동차'], negative=['반도체'])

[('기아차', 0.6842361092567444),
 ('기아', 0.564590573310852),
 ('위아', 0.545499324798584),
 ('전자', 0.5427474975585938),
 ('현대차', 0.5334987044334412),
 ('현대모비스', 0.5288388729095459),
 ('삼성전기', 0.5255287885665894),
 ('렉서스', 0.5211635828018188),
 ('아우디', 0.5193088054656982),
 ('제네시스', 0.5184187293052673)]

In [ ]:
model.wv.most_similar(positive=['하이닉스', '스마트폰'], negative=['반도체'])

[('삼성전자', 0.6542625427246094),
 ('갤럭시', 0.6359720826148987),
 ('아이폰', 0.5755826234817505),
 ('폴드', 0.56295245885849),
 ('아이패드', 0.5506658554077148),
 ('노트북', 0.5407217741012573),
 ('태블릿', 0.5389757752418518),
 ('삼성전기', 0.533003568649292),
 ('휴대폰', 0.5170853137969971),
 ('노트', 0.497602641582489)]